clang/lib/Basic/Diagnostic.cpp

```diff
@@ -477,6 +486,118 @@ void DiagnosticsEngine::setSeverityForAll(diag::Flavor Flavor,
       setSeverity(Diag, Map, Loc);
 }
 
+namespace {
+class WarningsSpecialCaseList : public llvm::SpecialCaseList {
+public:
+  static std::unique_ptr<WarningsSpecialCaseList>
+  create(const llvm::MemoryBuffer &MB, std::string &Err) {
+    auto SCL = std::make_unique<WarningsSpecialCaseList>();
+    if (!SCL->createInternal(&MB, Err))
+      return nullptr;
+    return SCL;
+  }
+
+  // Section names refer to diagnostic groups, which cover multiple individual
+  // diagnostics. Expand diagnostic groups here to individual diagnostics.
+  // A diagnostic can have multiple diagnostic groups associated with it, we let
+  // the last section take precedence in such cases.
+  void processSections(DiagnosticsEngine &Diags) {
+    // Drop the default section introduced by special case list, we only support
+    // exact diagnostic group names.
+    Sections.erase("*");
+    // Make sure we iterate sections by their line numbers.
+    std::vector<std::pair<unsigned, const llvm::StringMapEntry<Section> *>>
+        LineAndSectionEntry;
+    LineAndSectionEntry.reserve(Sections.size());
+    for (const auto &Entry : Sections) {
+      llvm::StringRef DiagName = Entry.first();
+      // Each section has a matcher with that section's name, attached to that
+      // line.
+      const auto &DiagSection = Entry.second.SectionMatcher;
+      unsigned DiagLine = DiagSection->Globs.at(DiagName).second;
+      LineAndSectionEntry.emplace_back(DiagLine, &Entry);
+    }
+    llvm::sort(LineAndSectionEntry);
+    static constexpr auto WarningFlavor = clang::diag::Flavor::WarningOrError;
+    for (const auto &[_, SectionEntry] : LineAndSectionEntry) {
+      SmallVector<diag::kind> GroupDiags;
+      llvm::StringRef DiagGroup = SectionEntry->getKey();
+      if (Diags.getDiagnosticIDs()->getDiagnosticsInGroup(
+              WarningFlavor, DiagGroup, GroupDiags)) {
+        StringRef Suggestion =
+            DiagnosticIDs::getNearestOption(WarningFlavor, DiagGroup);
+        Diags.Report(diag::warn_unknown_diag_option)
+            << static_cast<unsigned>(WarningFlavor) << DiagGroup
+            << !Suggestion.empty() << Suggestion;
+        continue;
+      }
+      for (diag::kind Diag : GroupDiags)
+        DiagToSection[Diag] = &SectionEntry->getValue();
```

Comment:
Consider adding a comment clarifying that we intentionally override previous values so that the latest entry will take precedence.